docs/src/docs/asciidoc/developer/clustered-events.adoc

```diff
@@ -0,0 +1,90 @@
+* Guaranteeing event delivery in all cases would require some form of stable store and a fairly complex and costly
+2-phase logic. This would also have an unsustainable performance impact. Instead, clients should have a way to figure out
+when such hazard happen to compensate for the possible lack of event delivery.
+
+== Technical facts
+
+Because of the current implementation:
+
+* Most events have to find their source from a client, more specifically from the clustered `Store` implementation:
+*expiry*, *removal*, *update* and *creation* are all client-triggered.
+* One event has to originate from the server: *eviction* is server-triggered.
+
+This means a cluster-wide listener mechanism has to be created with the following features:
+
+* Clients can register and unregister themselves.
+* Events can be fired from any client or any server.
+* Events have to be transported either from a server to all listening clients, or from one client to all clients
+(sometimes including the source client) via a server. AA topologies complicate client-sourced events transport as events
+have to be delivered only once.
+* Events can be fired from non-listening clients.
+
+== The transport mechanism
+
+An event delivery mechanism must be built to transport the events from the clients and servers to all clients registered
+as listeners. It requires the following:
+
+* API to (un)register a client as a listener
+* API for a client to fire an event
+* API for a server to fire an event
+* Implementation support of multiple servers
+
+A dedicated entity looks like the cleanest way ahead. This entity might need to be general-purpose enough to make
+it able to also transport invalidation events, and possibly any other type of notification logic between clients.
+Then refactor the invalidation mechanism to make use of this new notification entity. This should be regarded as a
+stretch goal once the implementation of this story is done.
+
+== The straightforward bits
+
+Assuming the event delivery mechanism is in place event firing logic for the following `Store` methods is going to be
+straightforward to implement as all the necessary data to decide what event to fire already is available handy:
+
+* putIfAbsent()
+* remove(1)
+* remove(2)
+* replace(2)
+* replace(3)
+
+The `Store` methods not listed above or below either throw `UnsupportedOperationException`, or delegate to other
+`Store` methods.
+
+== The complicated bits
+
+The following cases are going to be more complicated to implement:
+
+* All clients must be aware of the event firing logic, apply its rules and fire events even when some clients are not
+registered event listeners.
+* `Store.put()` needs to discriminate between *creation* and *update*. This means `resolve()` must be called, which also
+means that the current `append()` call must be replaced with the slower `getAndAppend()` one. This is going to put more
```

Comment:
If we move to a server initiated model then we only need call `append()`.  The initiating client is not special, it's listener will get fired when the server issues notifications that the chain was changed.

Reply:
This could work if the server wasn't a n-keys to operations store. That's going to cause some serious problems too if events are enabled while the clustered cache is being bulk-loaded, although this could be mitigated with extra work.